### PR TITLE
Feature/simon mini game

### PR DESCRIPTION
**Pull request by willy1989**

### CODE REVIEW

File: Assets/Simon/Scripts/ColorButtonEffectPlayer.cs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+            { Colors.Yellow, yellowButton.AudioClip },
+            { Colors.Blue, blueButton.AudioClip }
+        };
+
+        ToggleButtonHighlight(false);
+    }
+
+    private void ToggleButtonHighlight(bool onOff)
+    {
+        buttonHighlight.gameObject.SetActive(onOff);
```
Comment (maxwellpark) [inline Assets/Simon/Scripts/ColorButtonEffectPlayer.cs]: If all you are doing is flipping the bool, you can just do 
```csharp 
buttonHighlight.gameObject.SetActive(!buttonHighlight.gameObject.activeSelf); 
```

Don't need the parameter.

File: Assets/Simon/Scripts/GameLoopManager.cs [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+    private void StartGameLoop()
+    {
+        StartCoroutine(StartGameLoopCoroutine());
+    }
+
+    private IEnumerator StartGameLoopCoroutine()
+    {
+        score = 0;
+        UImanager.ToggleGameOverUI(onOff: false);
```
Comment (maxwellpark) [inline Assets/Simon/Scripts/GameLoopManager.cs]: uiManager is the MS convention, keeps it camelCase

File: Assets/Simon/Scripts/GameLoopManager.cs [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        if (currentMiniGamePhases == MiniGamePhases.WatchingPhase)
+        {
+            yield break;
+        }
+
+        currentMiniGamePhases = MiniGamePhases.WatchingPhase;
+        yield return StartCoroutine(colorButtonEffectPlayer.PlayButton(color));
+        currentMiniGamePhases = MiniGamePhases.PlayingPhase;
+
+        if (sequence.CompareColorWithCurrentSequenceColor(color) == false && currentMiniGamePhases != MiniGamePhases.WatchingPhase)
```
Comment (maxwellpark) [inline Assets/Simon/Scripts/GameLoopManager.cs]: You shouldn't need to check ```currentMiniGamePhases != MiniGamePhases.WatchingPhase``` here as you already did it on line 53 (it would have broken out already).

And you never see this ```== false``` these days. 

```!sequence.CompareColorWithCurrentSequenceColor(color)``` is considered correct, by MS etc.

File: Assets/Simon/Scripts/GameLoopManager.cs [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        else
+        {
+            sequence.IterateSequence();
+        }
+
+        if (sequence.SequenceReachedLastItem() == true)
+        {
+            currentMiniGamePhases = MiniGamePhases.WatchingPhase;
+            score++;
+            yield return new WaitForSeconds(1f);
```
Comment (maxwellpark) [inline Assets/Simon/Scripts/GameLoopManager.cs]: You could put the wait times in a constant or serialised variable if you wanted to easily reconfigure them later.

File: Assets/Simon/Scripts/Sequence.cs [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    {
+        return sequenceIndex >= ColorSequence.Count;
+    }
+
+    public void ResetSequenceIndex()
+    {
+        sequenceIndex = 0;
+    }
+
+    private Colors getRandomColor()
```
Comment (maxwellpark) [inline Assets/Simon/Scripts/Sequence.cs]: Pascal case for private functions too. Java is camelCase.

File: Assets/Simon/Scripts/UIManager.cs [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+public class UIManager : MonoBehaviour
+{
+    [SerializeField] private Button startGameButton;
+
+    [SerializeField] private Text gameOverText;
+
+    [SerializeField] private Text finalScoreText;
+
+    public void ToggleGameOverUI(bool onOff)
```
Comment (maxwellpark) [inline Assets/Simon/Scripts/UIManager.cs:14]: Could just flip them all instead

File: Assets/Simon/Scripts/UIManager.cs [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+    {
+        startGameButton.gameObject.SetActive(onOff);
+        gameOverText.gameObject.SetActive(onOff);
+        finalScoreText.gameObject.SetActive(onOff);
+    }
+
+    public void ChangeScore(int score)
+    {
+        finalScoreText.text = ""+score;
```
Comment (maxwellpark) [inline Assets/Simon/Scripts/UIManager.cs]: ```"" +``` looks a bit weird. 
Normally people would use ```score.ToString()```

Or ```$"{score}"```

File: Assets/Simon/Scripts/ColorButtonEffectPlayer.cs [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+
+    private AudioSource audioSource;
+
+    private Dictionary<Colors, Transform> colorButtonPositions;
+
+    private Dictionary<Colors, AudioClip> colorButtonSounds;
+
+    private void Awake()
+    {
+        buttonHighlight = GameObject.FindGameObjectWithTag("ButtonHighlight").transform;
```
Comment (maxwellpark) [inline Assets/Simon/Scripts/ColorButtonEffectPlayer.cs]: I guess this is a decent way of grabbing them without using the inspector. 

If it was a big game and you had a lot of tags it might be suboptimal though.

File: Assets/Simon/Scripts/ColorButton.cs [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+public class ColorButton : MonoBehaviour
+{
+    [SerializeField] private Colors color;
+
+    [SerializeField] private AudioClip _audioClip;
+
+    public AudioClip AudioClip
+    {
+        get
+        {
```
Comment (ambid17) [inline Assets/Simon/Scripts/ColorButton.cs]: public AudioClip AudioClip => _audioClip;

File: Assets/Simon/Scripts/ColorButtonEffectPlayer.cs (hunk 8 above)
Comment (ambid17) [inline Assets/Simon/Scripts/ColorButtonEffectPlayer.cs]: just set these references manually in the inspector next time. On small projects its worth saving the time. I'm not sure this prevents any sort of errors of poorly set up inspector

File: Assets/Simon/Scripts/Colors.cs [hunk 10]
```diff
@@ -0,0 +1,11 @@
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+public enum Colors
+{
+    Green,
```
Comment (ambid17) [inline Assets/Simon/Scripts/Colors.cs]: you can put this in another file that's more relevant, such as ColorButtonEffectPlayer

File: Assets/Simon/Scripts/GameLoopManager.cs [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+        UImanager = GameObject.FindObjectOfType<UIManager>();
+        colorButtonEffectPlayer = GameObject.FindObjectOfType<ColorButtonEffectPlayer>();
+    }
+
+    private void StartGameLoop()
+    {
+        StartCoroutine(StartGameLoopCoroutine());
+    }
+
+    private IEnumerator StartGameLoopCoroutine()
```
Comment (ambid17) [inline Assets/Simon/Scripts/GameLoopManager.cs:39]: this class feels unclean with the amount of IEnumerators, are they all necessary?

**Review by ambid17 — APPROVED** [review]
Some minor things to fix, but it's acceptable in its current state

File: Assets/Simon/Scripts/ColorButton.cs (hunk 9 above)
Comment (willy1989) [reply Assets/Simon/Scripts/ColorButton.cs]: Done. I just want to know why this is wrong. I did what I did to so that the audio clip can't be changed at runtime from another script. This is why there was no set for AudioClip. Is this correct?

File: Assets/Simon/Scripts/ColorButtonEffectPlayer.cs (hunk 8 above)
Comment (willy1989) [reply Assets/Simon/Scripts/ColorButtonEffectPlayer.cs]: OK done.

File: Assets/Simon/Scripts/Colors.cs (hunk 10 above)
Comment (willy1989) [reply Assets/Simon/Scripts/Colors.cs]: OK done. Is this a best practice to move enums to a class file? If you look for it in the project folders, isn't it easier to find if it is in its own file?

File: Assets/Simon/Scripts/GameLoopManager.cs (hunk 11 above)
Comment (willy1989) [reply Assets/Simon/Scripts/GameLoopManager.cs:39]: Actually, the game is split in phases. Before starting, each phase must wait for the previous one to end. This is why I use coroutines.

I don't know if coroutines are the best solution here, I just know that I need sequences to be played one after the other.

File: Assets/Simon/Scripts/GameLoopManager.cs [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+
+    private int score = 0;
+
+
+    private void Awake()
+    {
+        startGameButton.onClick.RemoveAllListeners();
+        startGameButton.onClick.AddListener(StartGameLoop);
+
+        sequence = GameObject.FindObjectOfType<Sequence>();
```
Comment (maxwellpark) [inline Assets/Simon/Scripts/GameLoopManager.cs]: ```GameObject.FindObjectOfType<T>``` can just be ```FindObjectOfType<T>``` as this class inherits from UnityEngine.Object

File: Assets/Simon/Scripts/Sequence.cs [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+            ExtendColorSequence();
+        }
+    }
+
+    public void IterateSequence()
+    {
+       sequenceIndex++;
+    }
+
+    public bool SequenceReachedLastItem()
```
Comment (maxwellpark) [inline Assets/Simon/Scripts/Sequence.cs]: I would use a property personally as it doesn't accept any parameters.

```public bool SequenceReachedLastItem => sequenceIndex >= ColorSequence.Count;```

File: Assets/Simon/Scripts/Sequence.cs [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    public void ResetSequenceIndex()
+    {
+        sequenceIndex = 0;
+    }
+
+    private Colors getRandomColor()
+    {
+        return (Colors)UnityEngine.Random.Range(0, 4);
```
Comment (maxwellpark) [inline Assets/Simon/Scripts/Sequence.cs]: If you change the indexes, or add more members (you have the max value hard-coded) to the enum, this will be invalid. 

I would do: 
```csharp        
        var allColors = Enum.GetValues(typeof(Colors));
        var randomColors = (Colors)allColors.GetValue(UnityEngine.Random.Range(0, allColors.Length));
```
Then it's always the correct range.

I can't remember if Unity random range is inclusive or exclusive, so it might have to be ```colors.Length - 1```

File: Assets/Simon/Scripts/Colors.cs [hunk 15]
```diff
@@ -0,0 +1,5 @@
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
```
Comment (maxwellpark) [inline Assets/Simon/Scripts/Colors.cs]: This file has only using directives? (no actual code)

File: Assets/Simon/Scripts/Sequence.cs [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+    public void CreateColorSequence(int sequenceLength)
+    {
+        if(sequenceLength <= 0)
+        {
+            Debug.LogError("The length of the sequence should be larger than 0");
+        }
+
+        ColorSequence = new List<Colors>();
+
+        for(int i = 0; i < sequenceLength; i++)
```
Comment (maxwellpark) [inline Assets/Simon/Scripts/Sequence.cs:23]: Instead of making a new List then using a for loop, you could use LINQ (functional paradigm).

```csharp

        ColorSequence = Enumerable.Range(0, sequenceLength - 1)
            .Select(x => getRandomColor()).ToList();
```

This might be slower than a for loop. I just like fluent interfaces (method chaining) lol.
https://en.wikipedia.org/wiki/Fluent_interface

File: Assets/Simon/Scripts/ColorButtonEffectPlayer.cs (hunk 1 above)
Comment (willy1989) [reply Assets/Simon/Scripts/ColorButtonEffectPlayer.cs]: Done

File: Assets/Simon/Scripts/GameLoopManager.cs (hunk 2 above)
Comment (willy1989) [reply Assets/Simon/Scripts/GameLoopManager.cs]: done

File: Assets/Simon/Scripts/GameLoopManager.cs (hunk 3 above)
Comment (willy1989) [reply Assets/Simon/Scripts/GameLoopManager.cs]: I removed currentMiniGamePhases != MiniGamePhases.WatchingPhase.

I still think adding "== false" makes thing clearer. I can remove it if you thing this is really weird.

File: Assets/Simon/Scripts/GameLoopManager.cs (hunk 4 above)
Comment (willy1989) [reply Assets/Simon/Scripts/GameLoopManager.cs]: Isn't creating a constant file just for this a bit overkill? For now I made it a serialized field.

File: Assets/Simon/Scripts/Sequence.cs (hunk 5 above)
Comment (willy1989) [reply Assets/Simon/Scripts/Sequence.cs]: My bad.

File: Assets/Simon/Scripts/UIManager.cs (hunk 6 above)
Comment (willy1989) [reply Assets/Simon/Scripts/UIManager.cs:14]: I don't think I can actually. Unless there is something I am missing. If I flip them, the game over UI will show at the wrong time. Besides, I like the level of control the onOff parameter provides. If the user of the interface just calls the function without it, he or she has no idea whether the UI will be turned on or off. It feels random and makes it hard to follow in my opinion.

File: Assets/Simon/Scripts/UIManager.cs (hunk 7 above)
Comment (willy1989) [reply Assets/Simon/Scripts/UIManager.cs]: done

File: Assets/Simon/Scripts/GameLoopManager.cs (hunk 12 above)
Comment (willy1989) [reply Assets/Simon/Scripts/GameLoopManager.cs]: done

File: Assets/Simon/Scripts/Sequence.cs (hunk 13 above)
Comment (willy1989) [reply Assets/Simon/Scripts/Sequence.cs]: Why is a property better than a method here? Because it is more compact?

File: Assets/Simon/Scripts/Sequence.cs (hunk 14 above)
Comment (willy1989) [reply Assets/Simon/Scripts/Sequence.cs]: Yes, I kept it very straightforward because the game is based on exactly 4 colors. But you are right, I'll do that.
By the way I am a bit confused about 
`var allColors = Enum.GetValues(typeof(Colors));`

What is it's return type. Because I tried to do the same thing without using var, but couldn't do it. 
I don't know much about var actually.

File: Assets/Simon/Scripts/Colors.cs (hunk 15 above)
Comment (willy1989) [reply Assets/Simon/Scripts/Colors.cs]: Done. I removed it.

File: Assets/Simon/Scripts/Sequence.cs (hunk 16 above)
Comment (willy1989) [reply Assets/Simon/Scripts/Sequence.cs:23]: I don't know anything about LINK yet. Is it okay if we keep things as they are now? I don't want to just copy paste something I don't understand.

File: Assets/Simon/Scripts/GameLoopManager.cs (hunk 4 above)
Comment (maxwellpark) [reply Assets/Simon/Scripts/GameLoopManager.cs]: It doesn't have to be in its own file. Just a constant variable in this class would be enough. 

Serialised field is good too.

File: Assets/Simon/Scripts/Sequence.cs (hunk 13 above)
Comment (maxwellpark) [reply Assets/Simon/Scripts/Sequence.cs]: I think because it's more like state than behaviour. That's my opinion anyway.

i.e. it's closer to data, rather than some operation. It has to do a comparison but to me it makes more sense if it reads like:

```if (SomeState) { } ```. rather than ```if(SomeOperation()) { }```

File: Assets/Simon/Scripts/Sequence.cs (hunk 14 above)
Comment (maxwellpark) [reply Assets/Simon/Scripts/Sequence.cs]: Yeah i just used var for an example, it is still strongly typed, but the compiler works out what type it is from the expression that is evaluated on that line. It is called 'implicit type' or 'inferred type'. It's just for local variables, you can't use it for a class member.

https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/implicitly-typed-local-variables

The return type would be ```Array```, which is the base class for all arrays in c#, so you have to cast it to the type of your enum. 

You could also cast the array rather than the element, which might be better than my example:

```csharp
Colors[] allColors = (Colors[])Enum.GetValues(typeof(Colors));
```
The Enum (base) class has some static methods like this so you can get the values or the names, which is nice.

File: Assets/Simon/Scripts/GameLoopManager.cs (hunk 3 above)
Comment (maxwellpark) [reply Assets/Simon/Scripts/GameLoopManager.cs]: Doesn't really bother me but it is not intuitive at all tbh.

Like how in formal logic, negation notation is always _not [something]_ , rather than _[something] not_

image

I think that's why it is the convention.

File: Assets/Simon/Scripts/Sequence.cs (hunk 16 above)
Comment (maxwellpark) [reply Assets/Simon/Scripts/Sequence.cs:23]: Yeah leave it as is. I just put it there if you were curious.

Comment (maxwellpark) [reply Assets/Simon/Scripts/Sequence.cs:23]: Also, never just copy paste code :D Should always test it and makes sure it works in the context of your app.

File: Assets/Simon/Scripts/ColorButton.cs (hunk 9 above)
Comment (maxwellpark) [reply Assets/Simon/Scripts/ColorButton.cs]:
```csharp
public AudioClip AudioClip => _audioClip;
```
This is readonly. There is no Set defined on the property.

It's just the modern C# way of doing things (called an 'expression-bodied member'): https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/statements-expressions-operators/expression-bodied-members. 


It is equivalent to:

```csharp
public AudioClip AudioClip { get { return _audioClip; } }
```

You don't actually need the separate serialised field too btw. You can do: 
```csharp 
[field: SerializeField] 
public AudioClip { get; private set; }
```
Which serialises the backing field on the property for you automatically.

Comment (willy1989) [reply Assets/Simon/Scripts/ColorButton.cs]: Ok got it.

File: Assets/Simon/Scripts/GameLoopManager.cs (hunk 3 above)
Comment (willy1989) [reply Assets/Simon/Scripts/GameLoopManager.cs]: I see. I'll give it a try then. In that case, I reworked the method to include "IS" in its name: ColorIsSameAsCurrentSequenceColor. I think the name is even clearer that way.

File: Assets/Simon/Scripts/GameLoopManager.cs (hunk 4 above)
Comment (willy1989) [reply Assets/Simon/Scripts/GameLoopManager.cs]: got it.

File: Assets/Simon/Scripts/Sequence.cs (hunk 13 above)
Comment (willy1989) [reply Assets/Simon/Scripts/Sequence.cs]: I see. It is a very simple comparison rather than an operation. No data is manipulated. There's no algorithm running.

File: Assets/Simon/Scripts/Sequence.cs (hunk 14 above)
Comment (willy1989) [reply Assets/Simon/Scripts/Sequence.cs]: I see. Actually I moved allColors = (Colors[])Enum.GetValues(typeof(Colors)) to a private field, since we only need get all the colors in the array only once. We don't need to call Enum.GetValues() every time we want to pick a random color.

`private Colors[] allColors = (Colors[])Enum.GetValues(typeof(Colors));`

What do you think?

Comment (maxwellpark) [reply Assets/Simon/Scripts/Sequence.cs]: Yeah looks good - doing it on one line is fine. I just did it on two for the sake of example.
